### PR TITLE
Introduce Puppet 4 datatypes and drop Puppet 3 support

### PR DESCRIPTION
**Pull request by mmerfort**
I have merged the PR #56 from @cyberious as good as I could.
I also had to add 2 of my own commits to make the tests and rubocop work.

### CODE REVIEW

File: manifests/conf.pp [hunk 1]
```diff
@@ -8,244 +8,102 @@
...
-    $size            = 'UNDEFINED',
-    $sharedscripts   = 'UNDEFINED',
-    $shred           = 'UNDEFINED',
-    $shredcycles     = 'UNDEFINED',
-    $start           = 'UNDEFINED',
-    $su_user         = 'UNDEFINED',
-    $su_group        = 'UNDEFINED',
-    $uncompresscmd   = 'UNDEFINED'
+  Stdlib::Unixpath $path = $name,
+  Enum['absent','present'] $ensure = 'present',
```
Comment (bastelfreak) [inline manifests/conf.pp]: please align those parameters at the =

File: examples/init.pp [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# The baseline for module testing used by Puppet Labs is that each manifest
+# should have a corresponding test manifest that declares that class or defined
```
Comment (bastelfreak) [inline examples/init.pp]: you can drop this file, because it doesn't contain any code

File: manifests/hourly.pp [hunk 3]
```diff
@@ -12,30 +12,27 @@
...
+    'absent'  => $ensure,
+    'present' => 'directory'
   }
 
-  file { "${logrotate::rules_configdir}/hourly":
-      ensure => $dir_ensure,
-      owner  => $logrotate::root_user,
-      group  => $logrotate::root_group,
-      mode   => '0755',
+  file { '/etc/logrotate.d/hourly':
```
Comment (robinbowes) [inline manifests/hourly.pp]: I think this hard-coded path will break on some distros. Any reason you changed this from `"${logrotate::rules_configdir}/hourly"` ?

File: manifests/defaults.pp [hunk 4]
```diff
@@ -1,172 +1,26 @@
...
-            create_group => 'root',
-        }
-      }
-    }
-    default: {
-      if !defined( Logrotate::Conf[$logrotate::logrotate_conf] ) {
-        logrotate::conf {$logrotate::logrotate_conf: }
+  $rules.each |$rule_name, $params| {
+    if !defined(Logrotate::Rule[$rule_name]) {
+      $_merged_params =
```
Comment (bastelfreak) [inline manifests/defaults.pp]: there is no need for a newline after the =

File: manifests/hourly.pp [hunk 5]
```diff
@@ -12,30 +12,27 @@
...
+    ensure => $dir_ensure,
+    owner  => 'root',
+    group  => 'root',
+    mode   => '0755',
   }
-  logrotate::cron { 'hourly':
+  file { '/etc/cron.hourly/logrotate':
```
Comment (robinbowes) [inline manifests/hourly.pp]: I think this hard-coded path will break on some distros

File: manifests/hourly.pp [hunk 6]
```diff
@@ -12,30 +12,27 @@
...
     ensure  => $ensure,
-    require => File["${logrotate::rules_configdir}/hourly"],
+    owner   => 'root',
+    group   => 'root',
+    mode    => '0555',
+    source  => 'puppet:///modules/logrotate/etc/cron.hourly/logrotate',
+    require => [ File['/etc/logrotate.d/hourly'], Package['logrotate'], ],
```
Comment (robinbowes) [inline manifests/hourly.pp]: I think this hard-coded path will break on some distros

File: manifests/rule.pp [hunk 7]
```diff
@@ -122,47 +122,52 @@
...
-                        $shred           = 'UNDEFINED',
-                        $shredcycles     = 'UNDEFINED',
-                        $start           = 'UNDEFINED',
-                        $su              = 'UNDEFINED',
-                        $su_owner        = 'UNDEFINED',
-                        $su_group        = 'UNDEFINED',
-                        $uncompresscmd   = 'UNDEFINED'
-                        ) {
+  Enum['present','absent'] $ensure   = 'present',
+  Optional[
```
Comment (bastelfreak) [inline manifests/rule.pp]: this is very very long. Should we put it in an own type?

File: manifests/params.pp [hunk 8]
```diff
@@ -1,21 +1,151 @@
...
     }
   }
-
-  $cron_daily_hour    = '1'
-  $cron_daily_minute  = '0'
-  $cron_hourly_minute = '01'
+  $cron_daily_hour    = 1
+  $cron_daily_minute  = 0
+  $cron_hourly_minute = 1
+  $config_file        = '/etc/logrotate.conf'
```
Comment (robinbowes) [inline manifests/params.pp]: I think this hard-coded path will break on some distros

File: manifests/rule.pp (hunk 7 above)
Comment (bastelfreak) [inline manifests/rule.pp]: same applies to the other multiline datatypes. We should put them in one line or in an own datatypes. This is what we did in the other modules as well.

File: metadata.json [hunk 9]
```diff
@@ -19,7 +20,7 @@
   "requirements": [
     {
       "name": "puppet",
-      "version_requirement": ">= 4.7.0 < 5.0.0"
+      "version_requirement": ">= 4.7.0 < 6.0.0"
```
Comment (bastelfreak) [inline metadata.json]: we need at least 4.9.0 here

Comment (mmerfort) [inline metadata.json]: The hiera thing isn't even used.
I'm not sure if I should stick to 4.7 and use the params.pp or use 4.9 and properly use hiera.

File: manifests/rule.pp (hunk 7 above)
Comment (mmerfort) [inline manifests/rule.pp]: There is a data type for ensurable in the module, I will use that and add some more.

File: metadata.json (hunk 9 above)
Comment (bastelfreak) [inline metadata.json]: IMO we should use it. The data is already present in this PR.

Comment (mmerfort) [inline metadata.json]: I'm not going to do this right now. I will make a new PR with more improvements which are unrelated to the original PR.

There are a lot of things to fix and I'm don't even think that the class structure of the module is any good.

**Comment by mmerfort** [comment]
I have updated the PR with the requested changes.
I'm going to change some more things in new PRs to make the Module ready for a new release.

**Comment by bastelfreak** [comment]
@robinbowes what do you think about the current state? IMO we should merge it as is.

**Comment by robinbowes** [comment]
I think those hard-coded paths need fixing before a merge.

**Comment by mmerfort** [comment]
Which hardcoded paths? I think the ones you mentioned have all been altered to allow changing.

**Comment by robinbowes** [comment]
Sorry -  I missed that you'd made those changes.

**Comment by xorpaul** [comment]
Did you intentionally remove the parameter `su` from the logrotate::rule?

The parameter still exists in the README.

I can open a merge request for this if you want

```
Optional[Boolean] $su                          = undef,
```

**Pull request merged**